cmd/crank/beta/diff/client/crossplane/composition_client_test.go

```diff
@@ -0,0 +1,1011 @@
+									"group": "example.org",
+									"names": map[string]interface{}{
+										"kind": "XExampleResource",
+									},
+									"claimNames": map[string]interface{}{
+										"kind": "ExampleResourceClaim",
+									},
+									"versions": []interface{}{
+										map[string]interface{}{
+											"name":          "v1",
+											"served":        true,
+											"referenceable": false, // No referenceable version
+										},
+										map[string]interface{}{
+											"name":          "v2",
+											"served":        true,
+											"referenceable": false, // No referenceable version
+										},
+									},
+								},
+							},
+						},
+					}, nil
+				}).
+				Build(),
+			fields: fields{
+				compositions: map[string]*apiextensionsv1.Composition{
+					"matching-comp": {
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "matching-comp",
+						},
+						Spec: apiextensionsv1.CompositionSpec{
+							CompositeTypeRef: apiextensionsv1.TypeReference{
+								APIVersion: "example.org/v1",
+								Kind:       "XExampleResource",
+							},
+						},
+					},
+				},
+			},
+			args: args{
+				ctx: context.Background(),
+				res: tu.NewResource("example.org/v1", "ExampleResourceClaim", "test-claim").
+					WithSpecField("compositionRef", map[string]interface{}{
+						"name": "matching-comp",
+					}).
+					Build(),
+			},
+			want: want{
+				composition: nil,
+				err:         errors.New("no referenceable version found in XRD"), // Should fail with this error
+			},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			// Create the CompositionClient
+			c := &DefaultCompositionClient{
+				resourceClient: &tt.mockResource,
+				logger:         tu.TestLogger(t, false),
+				compositions:   tt.fields.compositions,
+			}
+
+			// Test the FindMatchingComposition method
+			got, err := c.FindMatchingComposition(tt.args.ctx, tt.args.res)
+
+			if tt.want.err != nil {
+				if err == nil {
+					t.Errorf("\n%s\nFindMatchingComposition(...): expected error but got none", tt.reason)
+					return
+				}
+
+				if !strings.Contains(err.Error(), tt.want.err.Error()) {
+					t.Errorf("\n%s\nFindMatchingComposition(...): expected error containing %q, got %q",
+						tt.reason, tt.want.err.Error(), err.Error())
+				}
+				return
+			}
+
+			if err != nil {
+				t.Errorf("\n%s\nFindMatchingComposition(...): unexpected error: %v", tt.reason, err)
+				return
+			}
+
+			if tt.want.composition != nil {
+				if diff := cmp.Diff(tt.want.composition.Name, got.Name); diff != "" {
+					t.Errorf("\n%s\nFindMatchingComposition(...): -want composition name, +got composition name:\n%s", tt.reason, diff)
+				}
+
+				if diff := cmp.Diff(tt.want.composition.Spec.CompositeTypeRef, got.Spec.CompositeTypeRef); diff != "" {
+					t.Errorf("\n%s\nFindMatchingComposition(...): -want composition type ref, +got composition type ref:\n%s", tt.reason, diff)
+				}
+			}
+		})
+	}
```

Comment:
I suspect these test bodies could be simplified a bit, for example:

* You can use `cmp.EquateErrors` to test errors.
* I don't think all the `if tt.want.foo != nil` guards are needed - `cmp` should be able to handle the case where e.g. the returned error / composition is nil and that's what you expected.